modules/features2d/src/annoy.cpp

```diff
@@ -0,0 +1,265 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+#include "precomp.hpp"
+#include "../3rdparty/annoy/annoylib.h"
+#include <opencv2/core/utils/logger.hpp>
+
+
+namespace cv
+{
+
+struct Random
+{
+    static const uint64 default_seed = 0xffffffff;
+    #if __cplusplus < 201103L
+    typedef uint64 seed_type;
+    #endif
+
+    RNG rng;
+
+    Random(uint64 seed = default_seed)
+    {
+        rng.state = seed;
+    }
+
+    inline int flip()
+    {
+        // Draw random 0 or 1
+        return rng.next() & 1;
+    }
+
+    inline size_t index(size_t n)
+    {
+        // Draw random integer between 0 and n-1 where n is at most the number of data points you have
+        return rng(unsigned(n));
+    }
+
+    inline void set_seed(uint64 seed)
+    {
+        rng.state = seed;
+    }
+};
+
+template <typename DataType, typename DistanceType>
+class ANNIndexImpl : public ANNIndex
+{
+public:
+    ANNIndexImpl(int dimension) : dim(dimension)
+    {
+        index = makePtr<::cvannoy::AnnoyIndex<int, DataType, DistanceType, Random, ::cvannoy::AnnoyIndexSingleThreadedBuildPolicy>>(dimension);
+    }
+
+    void addItems(InputArray _dataset) CV_OVERRIDE
```

Comment:
I propose to return bool and print error to log, but not throw exception, if `addItems` was not successful.

Reply:
> I propose to return bool and print error to log, but not throw exception, if `addItems` was not successful.

Ok. 
And do you think it's better to return bool and print error too for `build` and `knnSearch`? My thought to throw exception was that there might be possible users who do not check the flags and just do addItems->build->knnSearch in a row which might cause crash.

Reply:
add_item returns false in a single case: if the state is loaded from file. So ignore my proposal, it's not relevant.